fix(service-worker): add `UNRECOVERABLE_STATE` handler

sonukapoor opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #36539


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

sonukapoor committed:
feat(service-worker): add `UnrecoverableStateError`

In several occasions it has been observed when the browser has evicted
eagerly cached assets from the cache and which can also not be found on the
server anymore. This can lead to broken state where only parts of the application
will load and others will fail.

This commit fixes this issue by checking for the missing asset in the cache
and on the server. If this condition is true, the broken client will be
notified about the current state through the `UnrecoverableStateError`.

Closes #36539

IgorMinar reviewed (commented):
I like where this is heading, but I suggest that we use terminology different from `unrecovered` because it gets awkward and confusing to use it. How about using `defunkt`, `terminalError`, or something of that sort?

gkalpak commented on packages/service-worker/src/low_level.ts:
```diff
@@ -33,6 +33,16 @@ export interface UpdateActivatedEvent {
 }
 
+/**
+ * An event emitted when a a resource can not be found anymore in the cache or on the server.
```
I would rather make this more generic, because in the future there might be other scenarios that constitute an unrecoverable state. How about something like:

```suggestion
 * An event emitted when the version of the app used by the service worker to serve this client is
 * in a broken state that cannot be recovered from and a full page reload is required.
 *
 * For example, this can happen when a required resource cannot be found neither in the cache nor on
 * the server.
```

gkalpak commented:
@IgorMinar, I was the one that proposed "unrecoverable"/"unrecoverableState" (I assume "unrecovered" is a typo). I find "unrecoverable" more descriptive than "defunkt" or "terminalError" (but not a hill I'm going to die on :grin:).

gkalpak reviewed (changes requested):
Thx, @sonukapoor! I've suggested some changes. LMK what you think.

Note to self: We need to document this in the appropriate guides (with code examples).

sonukapoor commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -147,10 +147,9 @@ export abstract class AssetGroup {
       // may specify things like credential inclusion, but for assets these are not honored in order
       // to avoid issues with opaque responses. The SW requests the data itself.
       const res = await this.fetchAndCacheOnce(this.adapter.newRequest(req.url));
-
```
Empty lines are haunting me 👻

sonukapoor commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -314,8 +313,12 @@ export abstract class AssetGroup {
       if (!res.ok) {
-        throw new Error(`Response not Ok (fetchAndCacheOnce): request for ${
-            req.url} returned response ${res.status} ${res.statusText}`);
+        const clients = await this.scope.clients.matchAll();
+        clients.forEach(client => {
+          client.postMessage({type: 'UNRECOVERABLE_STATE', url: req.url});
+        });
+
+        return res;
```
@gkalpak Any suggestions on where a better place would be to put that code?

gkalpak replied:
Off the top of my head, the `AssetGroup`'s `handleFetch()` method. It knows whether the URL was in `config.urls` (not in `patterns`), it knows whether it could be found in the cache or not, it knows that the server request failed (with an error code).

sonukapoor replied:
I assume you mean [here](https://github.com/angular/angular/blob/6e134b55af40d8d552b416577a496f305e890847/packages/service-worker/worker/src/assets.ts#L151): If I add the same code into that section, how do you identify which client is broken? I do not see anything on the `Client` that could help with that.

gkalpak replied:
Yes, I meant there. You are right that `AssetGroup` is not the right place to know which clients to notify. `AssetGroup` _is_ the right place to know whether we are in an unrecoverable state. So, it has to pass that information to the `Driver` (which knows which clients to notify.

Maybe, we could introduce a new `UnrecoverableStateError` error, which `AssetGroup` will throw in its `handleFetch()` method (if the conditions are met). Such an error will propagate through `AppVersion`'s [handleFetch()][1] method to `Driver`'s [handleFetch()][2] method.

The `Driver` can then check if it's an `UnrecoverableStateError` and handle it appropriately. It can determine the affected client similar to how we do it [here][3] and notify those clients only.

[1]: https://github.com/angular/angular/blob/8154bbd5381ace543941ac974adbf7f3f5ada717/packages/service-worker/worker/src/app-version.ts#L154
[2]: https://github.com/angular/angular/blob/8154bbd5381ace543941ac974adbf7f3f5ada717/packages/service-worker/worker/src/driver.ts#L436
[3]: https://github.com/angular/angular/blob/8154bbd5381ace543941ac974adbf7f3f5ada717/packages/service-worker/worker/src/driver.ts#L762-L764

sonukapoor replied:
That makes sense @gkalpak - I had added an additional commit to address - take a look and let me know what you think.

sonukapoor commented on packages/service-worker/src/update.ts:
```diff
@@ -31,6 +31,11 @@ export class SwUpdate {
   readonly activated: Observable<UpdateActivatedEvent>;
 
+  /**
+   * Emits an `UnrecoverableStateEvent` event whenever the resource can not longer be found in the cache or server.
```
You mean [this](https://github.com/angular/angular/pull/36847#discussion_r440904549)?

sonukapoor commented on packages/service-worker/src/low_level.ts:
```diff
@@ -33,6 +33,16 @@ export interface UpdateActivatedEvent {
+ */
+export interface UnrecoverableStateEvent {
+  type: 'UNRECOVERABLE_STATE';
+  url: string;
```
I assume you mean as an additional property on the interface?

gkalpak replied:
I mean replacing the `url` property of the interface with a `reason` property.

sonukapoor commented:
@gkalpak Please review again. I have made some of the suggested changes. I will address unit tests once we are OK with the changes.

gkalpak commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -150,6 +150,10 @@ export abstract class AssetGroup {
 
       // If this is successful, the response needs to be cloned as it might be used to respond to
       // multiple fetch operations at the same time.
```
This comment should be just above the closing operation.

gkalpak commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -150,6 +150,10 @@ export abstract class AssetGroup {
 
       // If this is successful, the response needs to be cloned as it might be used to respond to
       // multiple fetch operations at the same time.
+      if (!res.ok) {
```
AFAICT, `fetchAndCacheOnce()` will not return a not OK response. It will [throw an error][1] instead.

[1]: https://github.com/angular/angular/blob/5fa748b196d388f34aff2808bd8e4a7ab6b3af79/packages/service-worker/worker/src/assets.ts#L320-L323

gkalpak replied:
Also, as mentioned in https://github.com/angular/angular/pull/36847#discussion_r440821788 (and the subsequent discussion), the requirements for throwing an `UnrecoverableStateError` are:

- The request is matched by `config.urls` (not the `config.patterns`).
- It isn't found in the cache.
- It fails to be fetched by the server. (For now, we can detect this by checking whether `fetchAndCacheOnce()` failed with an non-critical error.)
- TBD: The `AssetGroup`'s `updateMode` should be `prefetch`, not `lazy`.

gkalpak commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -150,6 +150,10 @@ export abstract class AssetGroup {
       // If this is successful, the response needs to be cloned as it might be used to respond to
       // multiple fetch operations at the same time.
+      if (!res.ok) {
+        throw new UnrecoverableStateError(`Failed to retrieve asset from server`);
```
Please include the name of the `AssetGroup` and the resource URL in the reason.

gkalpak commented on packages/service-worker/worker/src/driver.ts:
```diff
@@ -1003,6 +1006,25 @@ export class Driver implements Debuggable, UpdateSource {
+
+    affectedClients.forEach(async clientId => {
+      const client = await this.scope.clients.get(clientId);
+      client.postMessage({type: 'UNRECOVERABLE_STATE', reason: 'Resource not found.'});
```
The reason should be passed through from the `UnrecoverableStateError`.

gkalpak reviewed (changes requested):
Thx for making the changes, @sonukapoor. Here is another round of review :smiley:

gkalpak commented on packages/service-worker/worker/src/error.ts:
```diff
@@ -17,3 +17,8 @@ export function errorToString(error: any): string {
   }
 }
+
+export class UnrecoverableStateError extends Error {
```
It just occurred to me that `SwUnrecoverableStateError` would be more consistent 😇

gkalpak commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -143,14 +143,27 @@ export abstract class AssetGroup {
+        const isInConfig = this.config.urls.indexOf(url) !== -1;
+        const isPrefetch = this.config.installMode === 'prefetch';
+
+        if (isCriticalError && isInConfig && isPrefetch) {
```
Actually, thinking about it some more, I don't think any error should trigger an `UnrecoverableStateError`. For example, the client of the server might be offline. We need to specifically detect 404 errors. Which means that we somehow need access to the raw `Response` object here.

Not sure what is the best way to achieve this. I have to investigate this further 😞

gkalpak replied:
So, here is a new plan:

- Update `cacheBustedFetchFromNetwrok()` to throw an `UnrecoverableStateError` if `cacheBustedResult.status` is 404:

    https://github.com/angular/angular/blob/1550663b9e5bcd7870380d870ef84aa5e9211957/packages/service-worker/worker/src/assets.ts#L415-L421

    Something like:
    ```ts
    if (!cacheBustedResult.ok) {
      if (cacheBustedResult.status === 404) {
        throw new UnrecoverabvleStateError(...);
      } else {
        throw new SwCriticalError(...);
      }
    } 
    ```

- Let the `Error` pass through in `handleFetch()`. At this point, if we get an `UnrecoverableStateError` we know that (1) the resource is missing from the cache, (2) the resource is missing from the server (404) and (3) the resource is hashed (which makes the `config.urls.indexOf()` check kind of redundant).

- Make `UnrecoverableStateError` extend `SwCriticalError`, because we still want to call [versionFailed()][1].

[1]: https://github.com/angular/angular/blob/ff8e66ca227017b9f97b67d8a7d17667a3b3d82b/packages/service-worker/worker/src/driver.ts#L754

gkalpak reviewed (changes requested):
Thx, @sonukapoor! Here are some more comments for you :grin:
(Some of my previous comments have not been addressed, so I left them unresolved.)

BTW, there have been some SW updates recently. Make sure you rebase this PR on latest master :wink:

sonukapoor commented on goldens/public-api/service-worker/service-worker.d.ts:
```diff
@@ -29,6 +29,7 @@ export declare class SwUpdate {
     readonly activated: Observable<UpdateActivatedEvent>;
     readonly available: Observable<UpdateAvailableEvent>;
     get isEnabled(): boolean;
+    readonly unrecovered: Observable<UnrecoverableStateEvent>;
```
The script that generates the `*.d.ts` file automatically changes the order of the properties.

aikidave commented:
Anyone know why this PR was assigned to me? :)

gkalpak commented:
Maybe because of [this comment][1]? 🤷‍♂️ 

[1]: https://github.com/angular/angular/pull/36847#pullrequestreview-430247130

gkalpak commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -422,10 +423,15 @@ export abstract class AssetGroup {
-                  cacheBustedResult.statusText}`);
+          if (cacheBustedResult.status === 404) {
+            throw new SwUnrecoverableStateError(
+                `Failed to retrieve asset from server: (AssetGroup): ${this.config.name}`);
```
I think it would be usefule to also capture the resource that failed to be retrieved. How about:

```suggestion
                `Failed to retrieve hashed resource from the server. (AssetGroup: ${this.config.name} | URL: ${url})`);
```

gkalpak reviewed (changes requested):
Looking much better (and simpler). I think we are almost there.
Now we need some tests and documentation :grin:

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,101 @@ describe('Driver', () => {
+                               .addFile('/bar.hash.js', 'console.log("BAR");')
+                               .build();
+
+      const originalManifest: Manifest = {
+        configVersion: 1,
+        timestamp: 1234567890123,
+        index: '/index.html',
+        assetGroups: [{
+          name: 'assets',
+          installMode: 'prefetch',
+          updateMode: 'prefetch',
+          urls: originalFiles.list(),
+          patterns: [],
+          cacheQueryOptions: {ignoreVary: true},
+        }],
+        dataGroups: [],
+        navigationUrls: processNavigationUrls(''),
+        hashTable: tmpHashTableForFs(originalFiles),
+      };
+
+      const updatedFilesManifest: Manifest = {
+        configVersion: 1,
+        timestamp: 1234567890123,
+        index: '/index.html',
+        assetGroups: [{
+          name: 'assets',
+          installMode: 'prefetch',
+          updateMode: 'prefetch',
+          urls: updatedFiles.list(),
+          patterns: [],
+          cacheQueryOptions: {ignoreVary: true},
+        }],
+        dataGroups: [],
+        navigationUrls: processNavigationUrls(''),
+        hashTable: tmpHashTableForFs(updatedFiles),
+      };
```
Let's create a helper function that accepts a `MockFileSystem` instance and generates a `Manifest`.
Or actually a helper that generates a `MockServerState` from a `MockFileSystem`.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,101 @@ describe('Driver', () => {
+      updatedServer.assertSawRequestFor('/foo.hash.js');
+
+      const mockClient = scope.clients.getMock('default')!;
+      expect(mockClient.messages).toEqual([
```
For the sake of completeness, let's assert that the client didn't have any messages before the last request for `/foo.hash.js`.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,101 @@ describe('Driver', () => {
+              'Failed to retrieve hashed resource from the server. (AssetGroup: assets | URL: /foo.hash.js)'
+        },
+      ]);
+    });
```
Can you also assert (potentially in a new, dedicated test) that the driver has triggered the `versionFailed()` sequence (e.g. entered degraded mode)?

gkalpak reviewed (changes requested):
Woohoo! Almost there :tada:

BTW, it occurred to me that this will not help much if the app has not loaded yet (i.e. if the error happens with one of the main JS bundles), but it _will_ help with lazy-loaded bundles (where the app has loaded and is capable of handling these messages).

We can introduce this first, see how it goes and later investigate more aggressive options (such as auto-reloading the client).

sonukapoor committed:
fix(service-worker): fix condition to check for a cache-busted request

Previously, the condition to make the cache busted was executing although
the network request was successful. However, this is not valid. The cache
should only be marked as busted when the request failed. This commit fixes
the invalid condition.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1762,6 +1875,27 @@ describe('Driver', () => {
 });
 })();
 
+function generateMockServerState(fileSystem: MockFileSystem) {
```
This helper is only used in two tests. I suggest wrapping these two tests in a `describe()` function and move the helper inside the `describe()` block.

E.g.:


```ts
describe('when in unrecoverable state', () => {
  const generateMockServerState = ...;

  it('does one thing', ...);
  
  it('does another thing', ...);
});
```

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,119 @@ describe('Driver', () => {
       expect(requestUrls2).toContain(httpsRequestUrl);
     });
 
+    it('broadcasts notifications about unrecoverable state', async () => {
```
Broadcasts is not accurate. It only notifies affected clients.

Now that I mentioned this, we should also have a test verifying that non-affected clients are not notified. For example:
- Create the server state with an original version.
- Get three different SW clients to the original version.
- Update the server to a new version (let's call it `version-2`).
- Update the first two clients to `version-2`. (Keep the third client to the original version.)
- Update the server to a newer version (let's call it `version-3`) and break the cache for `version-2` by removing an asset.
- Try to get the non-existent asset from the first client.
- Verify that the first two clients (which were on the broken version) were notified and that the third client (which was on a different version) was not notified.

gkalpak replied:
Actually, it probably makes more sense to update the current test to have three clients (as described above) instead of adding another test.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,119 @@ describe('Driver', () => {
+      // Remove `foo.hash.js` from the cache to emulate the browser evicting files from the cache.
+      await clearAllCaches(scope.caches);
+
+      const mockClient = scope.clients.getMock('default')!;
+      expect(mockClient.messages).toEqual([]);
```
These are irreelvant to this test. Please remove.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,119 @@ describe('Driver', () => {
+      expect(await makeRequest(scope, '/foo.hash.js')).toBeNull();
+      updatedServer.assertSawRequestFor('/foo.hash.js');
+
+      // This should also enter the `SW` into degrade mode.
```
```suggestion
      // This should also enter the `SW` into degraded mode, because the broken version was the
      // latest one.
```

gkalpak reviewed (changes requested):
Note to self: We still need to document this (with code snippets).

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
+        scope.updateServerState(latestServer);
+        expect(await driver.checkForUpdate()).toEqual(true);
+
+        // Remove `foo.hash.js` from the cache to emulate the browser evicting files from the cache.
```
As discussed offline, this is not ideal (removing all caches instead of just removing `foo.hash.js` from the cache), but is good enough for our purposes here. Please add a comment explaining that. For example:

```suggestion
        // Remove `foo.hash.js` from the cache to emulate the browser evicting files from the cache.
        // NOTE: For simplicity, we are clearing all caches instead of just removing `foo.hash.js`.
        //       This is not ideal, but OK for our purposes here.
```

(Same for similar lines in other tests in this `describe()` block.)

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
+
+        // Get first client and verify its messages
+        const mockClient1 = scope.clients.getMock('client1')!;
+        // Clear all messages for this client
```
What messages do we expect this client to have?

sonukapoor replied:
Because we are updating the server state, the following messages in that array: `UPDATE_AVAILABLE`, `UPDATE_ACTIVATED`, `UPDATE_AVAILABLE`. I am clearing those first because I want to validate the `UNRECOVERABLE_STATE`.

gkalpak replied:
:+1:
Can you update the comment to mention that. Something like:

```suggestion
        // Clear messages related to previous updates for this client
```

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
+      };
+
+      it('notify affected clients about unrecoverable state', async () => {
+        const originalServer = generateMockServerState(
```
IMO, it would be clearer to use numbers for server version names. I.e. instead of `originalServer`, `updatedServer`, `latestServer` use `serverState1`, `serverState2`, `serverState3`.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
+          },
+        ]);
+
+        // Because `client` failed, `client2` has been moved to the latest version.
```
```suggestion
        // Because `client1` failed, `client2` has been moved to the latest version.
```

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
+            reason:
+                'Failed to retrieve hashed resource from the server. (AssetGroup: assets | URL: /bar.hash.js)'
+          },
+        ]);
```
We should verify the same for `client2` and also verify that `client3` did not receive any messages.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
+
+        // Because `client` failed, `client2` has been moved to the latest version.
+        // Verify that by retrieving `baz.hash.js`.
+        expect(await makeRequest(scope, '/baz.hash.js', 'client2')).toBe('console.log("BAZ");');
```
We should verify the same for `client1`.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
+        // Verify that by retrieving `baz.hash.js`.
+        expect(await makeRequest(scope, '/baz.hash.js', 'client2')).toBe('console.log("BAZ");');
+        updatedServer.assertNoRequestFor('/baz.hash.js');
+      });
```
We should also verify that `client3` remains on the first version (and can successfully request `foo.hash.js`).

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
 
-      const mockClient = scope.clients.getMock('default')!;
-      expect(mockClient.messages).toEqual([]);
+        const mockClient = scope.clients.getMock('default')!;
```
Please remove this too.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
+            .build();
+      };
+
+      it('notify affected clients about unrecoverable state', async () => {
```
```suggestion
      it('notifies affected clients', async () => {
```

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,117 +1738,161 @@ describe('Driver', () => {
 
-      const originalServer = generateMockServerState(originalFiles);
-      const updatedServer = generateMockServerState(updatedFiles);
+      it('enters degraded mode when unrecoverable state', async () => {
```
```suggestion
      it('enters degraded mode', async () => {
```

sonukapoor committed:
test(service-worker): add helper function remove individual cache

This commit adds a helper method to remove individual cached items.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,170 @@ describe('Driver', () => {
+                           .withManifest(manifest)
+                           .withStaticFiles(fileSystem)
+                           .build(),
+          manifest
```
```suggestion
          manifest,
```

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,170 @@ describe('Driver', () => {
+                .addFile('/index.html', '<script src="bar.hash.js"></script>')
+                .addFile('/bar.hash.js', 'console.log("BAR");')
+                .build());
+
```
Let's be consistent with the creation of server states: Either create all server states (including `serverState3`) at the top or create each state right before it is used.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,170 @@ describe('Driver', () => {
+        expect(await makeRequest(scope, '/foo.hash.js')).toBe('console.log("FOO");');
+        serverState1.assertNoRequestFor('/foo.hash.js');
+
+        // Update the server to the updated version.
```
```suggestion
        // Update the ServiceWorker to the second version.
```

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,170 @@ describe('Driver', () => {
+            reason:
+                'Failed to retrieve hashed resource from the server. (AssetGroup: assets | URL: /bar.hash.js)'
+          },
+        ]);
```
We should verify the same for `client2` and also verify that `client3` did not receive any messages.

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1738,6 +1738,170 @@ describe('Driver', () => {
+                    .build();
+        driver = new Driver(scope, scope, new CacheDatabase(scope, scope));
+
+
```
```suggestion
```

gkalpak reviewed (changes requested):
Final(?) round of review comments :smiley:

Please squash the existing fixup commits before making the new changes. Also, the commit that adds the test helper should have a type of `test` (`test(service-worker): ...`). BTW, the test helper cannot be used in a commit before the commit that adds it :wink:

The final bit is documenting this in the [ServiceWorker communications][1] guide.

[1]: https://angular.io/guide/service-worker-communications

sonukapoor committed:
fixup! feat(service-worker): add `UnrecoverableStateError`

sonukapoor committed:
docs(service-worker): add section to explain unrecoverable state

sonukapoor commented:
> ...BTW, the test helper cannot be used in a commit before the commit that adds it 😉

😁 Yes, of course.

mary-poppins commented:
You can preview 1d2fa37 at https://pr36847-1d2fa37.ngbuilds.io/.
You can preview 6f833e7 at https://pr36847-6f833e7.ngbuilds.io/.
You can preview 9d1f214 at https://pr36847-9d1f214.ngbuilds.io/.
You can preview fd1ab83 at https://pr36847-fd1ab83.ngbuilds.io/.
You can preview 9843c4f at https://pr36847-9843c4f.ngbuilds.io/.
You can preview 6e134b5 at https://pr36847-6e134b5.ngbuilds.io/.
You can preview 5fa748b at https://pr36847-5fa748b.ngbuilds.io/.
You can preview 6450542 at https://pr36847-6450542.ngbuilds.io/.
You can preview ff8e66c at https://pr36847-ff8e66c.ngbuilds.io/.
You can preview 919ff58 at https://pr36847-919ff58.ngbuilds.io/.
You can preview 7d930ac at https://pr36847-7d930ac.ngbuilds.io/.
You can preview 81f0b81 at https://pr36847-81f0b81.ngbuilds.io/.
You can preview 3a7ed6b at https://pr36847-3a7ed6b.ngbuilds.io/.
You can preview 3f12164 at https://pr36847-3f12164.ngbuilds.io/.
You can preview 82f7d91 at https://pr36847-82f7d91.ngbuilds.io/.
You can preview 7c2e122 at https://pr36847-7c2e122.ngbuilds.io/.
You can preview 94113f4 at https://pr36847-94113f4.ngbuilds.io/.
You can preview be202f6 at https://pr36847-be202f6.ngbuilds.io/.
You can preview 7715875 at https://pr36847-7715875.ngbuilds.io/.

gkalpak reviewed (changes requested):
This is almost ready :tada:
Great job, @sonukapoor 🏆 

FYI, I tried it with angular.io (by manually removing one of the lazy bundles from both the cache and the server and everything seems to work fine 💯).

sonukapoor committed:
fixup! docs(service-worker): add section to explain unrecoverable state

gkalpak commented on aio/content/examples/service-worker-getting-started/src/app/unrecoverable-state.service.ts:
```diff
@@ -0,0 +1,14 @@
+
+  constructor(updates: SwUpdate) {
+    updates.unrecoverable.subscribe(event => {
+        document.location.reload();
```
Yes, sorry 😇

gkalpak commented on aio/content/guide/service-worker-communications.md:
```diff
@@ -67,6 +67,24 @@ Therefore, it is recommended to reload the page once the promise returned by `ac
+- At some later point, the app requests the lazy bundle, `lazy-chunk.<lazy-hash-1>.js`.
+- The service worker is unable to find the asset in the cache (remember that the browser evicted it). Nor is it able to retrieve it from the server (since the server now only has `lazy-chunk.<lazy-hash-2>.js` from the newer version).
+
+In the above scenario, the service worker is not able to serve an asset that would normally be cached. That particular app version is broken and there is no way to fix it without reloading the page. In such cases, the service worker notifies the client by sending a `UnrecoverableStateEvent` event. You can subscribe to `SwUpdate#unrecoverable` to be notified and handle these errors.
```
```suggestion
In the above scenario, the service worker is not able to serve an asset that would normally be cached. That particular app version is broken and there is no way to fix it without reloading the page. In such cases, the service worker notifies the client by sending an `UnrecoverableStateEvent` event. You can subscribe to `SwUpdate#unrecoverable` to be notified and handle these errors.
```

Also, reading this again I think it would be a little more clear to say something like:

```suggestion
In the above scenario, the service worker is not able to serve an asset that would normally be cached. That particular app version is broken and there is no way to fix the state of the client without reloading the page. In such cases, the service worker notifies the client by sending a `UnrecoverableStateEvent` event. You can subscribe to `SwUpdate#unrecoverable` to be notified and handle these errors.
```

gkalpak commented on aio/content/guide/service-worker-communications.md:
```diff
@@ -67,6 +67,24 @@ Therefore, it is recommended to reload the page once the promise returned by `ac
+In some cases, the version of the app used by the service worker to serve a client might be in a broken state that cannot be recovered from without a full page reload.
+
+For example, imagine the following scenario:
+- A user opens the app for the first time and the service worker caches the latest version of the app. Let's assume the app's cached assets include `index.html`, `main.<main-hash-1>.js` and `lazy-chunk.<lazy-hash-1>.js`.
```
I just remembered we are trying to standardize on having one sentence per line in Markdown files. Can you please ensure new sentences are placed on a new line in this section?

```suggestion
- A user opens the app for the first time and the service worker caches the latest version of the app.
  Let's assume the app's cached assets include `index.html`, `main.<main-hash-1>.js` and `lazy-chunk.<lazy-hash-1>.js`.
```

gkalpak commented on aio/content/examples/service-worker-getting-started/src/app/unrecoverable-state.service.ts:
```diff
@@ -0,0 +1,20 @@
+
+// #docregion sw-unrecoverable-state
+@Injectable()
+export class HandleUnrecoverableStateService {
```
We should also rename the file to match the service name:

`unrecoverable-state.service.ts` --> `handle-unrecoverable-state.service.ts`

gkalpak reviewed (changes requested):
Some minor comments. Otherwise lgtm :tada:

sonukapoor committed:
fixup! docs(service-worker): add section to explain unrecoverable state

mary-poppins commented:
You can preview ec812b8 at https://pr36847-ec812b8.ngbuilds.io/.

gkalpak commented on aio/content/guide/service-worker-communications.md:
```diff
@@ -67,6 +67,33 @@ Therefore, it is recommended to reload the page once the promise returned by `ac
+
+In the above scenario, the service worker is not able to serve an asset that would normally be cached.
+That particular app version is broken and there is no way to fix the state of the client without reloading the page.
+In such cases, the service worker notifies the client by sending a `UnrecoverableStateEvent` event.
```
```suggestion
In such cases, the service worker notifies the client by sending an `UnrecoverableStateEvent` event.
```

gkalpak commented on aio/content/guide/service-worker-communications.md:
```diff
@@ -67,6 +67,33 @@ Therefore, it is recommended to reload the page once the promise returned by `ac
+
+For example, imagine the following scenario:
+- A user opens the app for the first time and the service worker caches the latest version of the app.
+Let's assume the app's cached assets include `index.html`, `main.<main-hash-1>.js` and `lazy-chunk.<lazy-hash-1>.js`.
```
```suggestion
  Let's assume the app's cached assets include `index.html`, `main.<main-hash-1>.js` and `lazy-chunk.<lazy-hash-1>.js`.
```

(Below too.)

sonukapoor committed:
fixup! docs(service-worker): add section to explain unrecoverable state

mary-poppins commented:
You can preview a363e95 at https://pr36847-a363e95.ngbuilds.io/.

gkalpak commented on packages/service-worker/src/low_level.ts:
```diff
@@ -33,6 +33,22 @@ export interface UpdateActivatedEvent {
+ * cache nor from the server. This could happen if a new version is deployed to the server and the
+ * ServiceWorker cache has been partially cleaned by the browser, removing some files of a previous
+ * app version but not all.
+ *
```
I think it is worth adding a link to the [Service worker communication][1] guide (here and in the other events too). Something like:

```suggestion
 *
 * @see {@link guide/service-worker-communications Service worker communication guide}
 *
```

And it's also a good idea to add a similar annotation in the `SwUpdate` class jsdoc.
(Adding these links is a separate change, so feel free to make them on a new commit (either on this PR or a new one).)

[1]: https://angular.io/guide/service-worker-communications

sonukapoor committed:
docs(service-worker): add links to service worker communication guide

sonukapoor committed:
fixup! feat(service-worker): add `UnrecoverableStateError`

mary-poppins commented:
You can preview 01a67a7 at https://pr36847-01a67a7.ngbuilds.io/.

gkalpak reviewed (approved):
LGTM :tada:

Great job, @sonukapoor 🏆
I know this wasn't a walk in the park and it took a few iterations, but the end result is quite clean and well worth it. Thank you for seeing it through :smiley:

BTW, we also need a public API review (~~not sure why PullApprove didn't request one already~~ _it did just now_ 🤷‍♂️).

sonukapoor commented:
> LGTM 🎉
> 
> Great job, @sonukapoor 🏆
> I know this wasn't a walk in the park and it took a few iterations, but the end result is quite clean and well worth it. Thank you for seeing it through 😃

I definitely agree with that. It felt more like a hike in the Himalayas... 😄 
Thank you though for guiding me through this. 
I would have not been complete this without you. 
I have learned a lot from this PR.

atscott reviewed (approved):
I haven't looked at the whole progression of the conversation here, but this final state looks really well documented and clear!

LGTM
Reviewed-for: public-api

> BTW, we also need a public API review (~not sure why PullApprove didn't request one already~ _it did just now_ 🤷‍♂️).

PullApprove now waits until initial approvals are there before requesting public API review to reduce noise to that group :)

petebacondarwin commented on packages/service-worker/test/comm_spec.ts:
```diff
@@ -435,6 +435,14 @@ import {MockPushManager, MockPushSubscription, MockServiceWorkerContainer, MockS
           },
         });
       });
+      it('processes resource removed notifications when sent', done => {
```
```suggestion
      it('processes unrecoverable notifications when sent', done => {
```

petebacondarwin reviewed (approved):
Reviewed-for: public-api

sonukapoor committed:
fixup! feat(service-worker): add `UnrecoverableStateError`

mary-poppins commented:
You can preview ab76f91 at https://pr36847-ab76f91.ngbuilds.io/.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

govi2010 commented:
How can I get this fix? 
I am following this PR which is closed. 
I am using PWA with angular 9. 
How can I get this Fix ?

gkalpak commented:
@govi2010, this feature was introduced in v11.0.0-next.0. The first stable version that will have it is 11.0.0.
